pootle/apps/pootle_fs/plugin.py

```diff
@@ -0,0 +1,344 @@
+                if pootle_wins
+                else "staged_for_merge_fs")
+            response.add(action_type, fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    def rm(self, state, response, fs_path=None, pootle_path=None, force=False):
+        """
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        """
+        removed = (
+            state["pootle_removed"]
+            + state["fs_removed"]
+            + state["both_removed"])
+        if force:
+            removed = (
+                removed
+                + state["fs_untracked"]
+                + state["pootle_untracked"])
+        for fs_state in removed:
+            if fs_state.state_type in ["fs_untracked", "pootle_untracked"]:
+                fs_state.kwargs["store_fs"] = self.create_store_fs(
+                    pootle_path=fs_state.pootle_path,
+                    fs_path=fs_state.fs_path)
+            fs_state.store_fs.file.rm()
+            response.add("staged_for_removal", fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    def sync_merge(self, state, response, fs_path=None, pootle_path=None):
+        for fs_state in (state["merge_pootle_wins"] + state["merge_fs_wins"]):
+            pootle_wins = (fs_state.state_type == "merge_pootle_wins")
+            response_type = (
+                "merged_from_pootle"
+                if pootle_wins
+                else "merged_from_fs")
+            store_fs = fs_state.store_fs
+            store_fs.file.pull(
+                merge=True,
+                pootle_wins=pootle_wins,
+                user=self.pootle_user)
+            store_fs.file.push()
+            store_fs.file.on_sync()
+            response.add(
+                response_type,
+                fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    @emits_state(pre=fs_pre_pull, post=fs_post_pull)
+    def sync_pull(self, state, response, fs_path=None, pootle_path=None):
+        """
+        :param fs_path: Path glob to filter translations matching FS path
+        :param pootle_path: Path glob to filter translations to add matching
+          ``pootle_path``
+        """
+        for fs_state in (state['fs_staged'] + state['fs_ahead']):
+            fs_state.store_fs.file.pull(user=self.pootle_user)
+            fs_state.store_fs.file.on_sync()
+            response.add("pulled_to_pootle", fs_state=fs_state)
+        return response
+
+    @responds_to_state
+    @emits_state(pre=fs_pre_push, post=fs_post_push)
+    def sync_push(self, state, response, fs_path=None, pootle_path=None):
+        pushable = state['pootle_staged'] + state['pootle_ahead']
+        for fs_state in pushable:
+            fs_state.store_fs.file.push()
+            response.add('pushed_to_fs', fs_state=fs_state)
+        self.push(response)
```

Comment:
tbh its more about where plugins (git, mercurial etc) need to intersect, and how to properly handle failure